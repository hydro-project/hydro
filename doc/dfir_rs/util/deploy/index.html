<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="Hydro Deploy integration for DFIR."><title>dfir_rs::util::deploy - Rust</title><script>if(window.location.protocol!=="file:")document.head.insertAdjacentHTML("beforeend","SourceSerif4-Regular-6b053e98.ttf.woff2,FiraSans-Italic-81dc35de.woff2,FiraSans-Regular-0fe48ade.woff2,FiraSans-MediumItalic-ccf7e434.woff2,FiraSans-Medium-e1aa3f0a.woff2,SourceCodePro-Regular-8badfe75.ttf.woff2,SourceCodePro-Semibold-aa29a496.ttf.woff2".split(",").map(f=>`<link rel="preload" as="font" type="font/woff2"href="../../../static.files/${f}">`).join(""))</script><link rel="stylesheet" href="../../../static.files/normalize-9960930a.css"><link rel="stylesheet" href="../../../static.files/rustdoc-e56847b5.css"><meta name="rustdoc-vars" data-root-path="../../../" data-static-root-path="../../../static.files/" data-current-crate="dfir_rs" data-themes="" data-resource-suffix="" data-rustdoc-version="1.92.0-nightly (f6092f224 2025-09-22)" data-channel="nightly" data-search-js="search-e256b49e.js" data-stringdex-js="stringdex-061df703.js" data-settings-js="settings-c38705f0.js" ><script src="../../../static.files/storage-e2aeef58.js"></script><script defer src="../sidebar-items.js"></script><script defer src="../../../static.files/main-6dc2a7f3.js"></script><noscript><link rel="stylesheet" href="../../../static.files/noscript-263c88ec.css"></noscript><link rel="alternate icon" type="image/png" href="../../../static.files/favicon-32x32-eab170b8.png"><link rel="icon" type="image/svg+xml" href="../../../static.files/favicon-044be391.svg"></head><body class="rustdoc mod"><!--[if lte IE 11]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><rustdoc-topbar><h2><a href="#">Module deploy</a></h2></rustdoc-topbar><nav class="sidebar"><div class="sidebar-crate"><h2><a href="../../../dfir_rs/index.html">dfir_rs</a><span class="version">0.14.0</span></h2></div><div class="sidebar-elems"><section id="rustdoc-toc"><h2 class="location"><a href="#">Module deploy</a></h2><h3><a href="#modules">Module Items</a></h3><ul class="block"><li><a href="#modules" title="Modules">Modules</a></li><li><a href="#macros" title="Macros">Macros</a></li><li><a href="#structs" title="Structs">Structs</a></li><li><a href="#enums" title="Enums">Enums</a></li><li><a href="#traits" title="Traits">Traits</a></li><li><a href="#functions" title="Functions">Functions</a></li><li><a href="#types" title="Type Aliases">Type Aliases</a></li></ul></section><div id="rustdoc-modnav"><h2><a href="../index.html">In dfir_<wbr>rs::<wbr>util</a></h2></div></div></nav><div class="sidebar-resizer" title="Drag to resize sidebar"></div><main><div class="width-limiter"><section id="main-content" class="content"><div class="main-heading"><div class="rustdoc-breadcrumbs"><a href="../../index.html">dfir_rs</a>::<wbr><a href="../index.html">util</a></div><h1>Module <span>deploy</span>&nbsp;<button id="copy-path" title="Copy item path to clipboard">Copy item path</button></h1><rustdoc-toolbar></rustdoc-toolbar><span class="sub-heading"><a class="src" href="../../../src/dfir_rs/util/deploy.rs.html#1-116">Source</a> </span></div><span class="item-info"><div class="stab portability">Available on <strong>crate feature <code>deploy_integration</code></strong> only.</div></span><details class="toggle top-doc" open><summary class="hideme"><span>Expand description</span></summary><div class="docblock"><p>Hydro Deploy integration for DFIR.</p>
</div></details><h2 id="modules" class="section-header">Modules<a href="#modules" class="anchor">§</a></h2><dl class="item-table"><dt><a class="mod" href="multi_connection/index.html" title="mod dfir_rs::util::deploy::multi_connection">multi_<wbr>connection</a></dt></dl><h2 id="macros" class="section-header">Macros<a href="#macros" class="anchor">§</a></h2><dl class="item-table"><dt><a class="macro" href="macro.launch.html" title="macro dfir_rs::util::deploy::launch">launch</a></dt></dl><h2 id="structs" class="section-header">Structs<a href="#structs" class="anchor">§</a></h2><dl class="item-table"><dt><a class="struct" href="struct.ConnectedDemux.html" title="struct dfir_rs::util::deploy::ConnectedDemux">Connected<wbr>Demux</a></dt><dt><a class="struct" href="struct.ConnectedDirect.html" title="struct dfir_rs::util::deploy::ConnectedDirect">Connected<wbr>Direct</a></dt><dt><a class="struct" href="struct.ConnectedTagged.html" title="struct dfir_rs::util::deploy::ConnectedTagged">Connected<wbr>Tagged</a></dt><dt><a class="struct" href="struct.DemuxDrain.html" title="struct dfir_rs::util::deploy::DemuxDrain">Demux<wbr>Drain</a></dt><dt><a class="struct" href="struct.DeployPorts.html" title="struct dfir_rs::util::deploy::DeployPorts">Deploy<wbr>Ports</a></dt><dd>Contains runtime information passed by Hydro Deploy to a program,
describing how to connect to other services and metadata about them.</dd><dt><a class="struct" href="struct.MergeSource.html" title="struct dfir_rs::util::deploy::MergeSource">Merge<wbr>Source</a></dt><dt><a class="struct" href="struct.TaggedSource.html" title="struct dfir_rs::util::deploy::TaggedSource">Tagged<wbr>Source</a></dt></dl><h2 id="enums" class="section-header">Enums<a href="#enums" class="anchor">§</a></h2><dl class="item-table"><dt><a class="enum" href="enum.AcceptedServer.html" title="enum dfir_rs::util::deploy::AcceptedServer">Accepted<wbr>Server</a></dt><dt><a class="enum" href="enum.BoundServer.html" title="enum dfir_rs::util::deploy::BoundServer">Bound<wbr>Server</a></dt><dt><a class="enum" href="enum.ClientConnection.html" title="enum dfir_rs::util::deploy::ClientConnection">Client<wbr>Connection</a></dt><dt><a class="enum" href="enum.Connection.html" title="enum dfir_rs::util::deploy::Connection">Connection</a></dt><dt><a class="enum" href="enum.ServerBindConfig.html" title="enum dfir_rs::util::deploy::ServerBindConfig">Server<wbr>Bind<wbr>Config</a></dt><dt><a class="enum" href="enum.ServerPort.html" title="enum dfir_rs::util::deploy::ServerPort">Server<wbr>Port</a></dt><dd>Describes how to connect to a service which is listening on some port.</dd></dl><h2 id="traits" class="section-header">Traits<a href="#traits" class="anchor">§</a></h2><dl class="item-table"><dt><a class="trait" href="trait.Connected.html" title="trait dfir_rs::util::deploy::Connected">Connected</a></dt><dt><a class="trait" href="trait.ConnectedSink.html" title="trait dfir_rs::util::deploy::ConnectedSink">Connected<wbr>Sink</a></dt><dt><a class="trait" href="trait.ConnectedSource.html" title="trait dfir_rs::util::deploy::ConnectedSource">Connected<wbr>Source</a></dt><dt><a class="trait" href="trait.StreamSink.html" title="trait dfir_rs::util::deploy::StreamSink">Stream<wbr>Sink</a></dt></dl><h2 id="functions" class="section-header">Functions<a href="#functions" class="anchor">§</a></h2><dl class="item-table"><dt><a class="fn" href="fn.accept_bound.html" title="fn dfir_rs::util::deploy::accept_bound">accept_<wbr>bound</a></dt><dt><a class="fn" href="fn.init.html" title="fn dfir_rs::util::deploy::init">init</a></dt><dt><a class="fn" href="fn.init_no_ack_start.html" title="fn dfir_rs::util::deploy::init_no_ack_start">init_<wbr>no_<wbr>ack_<wbr>start</a></dt><dt><a class="fn" href="fn.launch_flow.html" title="fn dfir_rs::util::deploy::launch_flow">launch_<wbr>flow</a></dt></dl><h2 id="types" class="section-header">Type Aliases<a href="#types" class="anchor">§</a></h2><dl class="item-table"><dt><a class="type" href="type.BufferedDrain.html" title="type dfir_rs::util::deploy::BufferedDrain">Buffered<wbr>Drain</a></dt><dt><a class="type" href="type.DynSink.html" title="type dfir_rs::util::deploy::DynSink">DynSink</a></dt><dt><a class="type" href="type.DynStream.html" title="type dfir_rs::util::deploy::DynStream">DynStream</a></dt><dt><a class="type" href="type.DynStreamSink.html" title="type dfir_rs::util::deploy::DynStreamSink">DynStream<wbr>Sink</a></dt><dt><a class="type" href="type.InitConfig.html" title="type dfir_rs::util::deploy::InitConfig">Init<wbr>Config</a></dt></dl></section></div></main></body></html>