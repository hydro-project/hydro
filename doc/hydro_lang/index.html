<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="API documentation for the Rust `hydro_lang` crate."><title>hydro_lang - Rust</title><script>if(window.location.protocol!=="file:")document.head.insertAdjacentHTML("beforeend","SourceSerif4-Regular-46f98efaafac5295.ttf.woff2,FiraSans-Regular-018c141bf0843ffd.woff2,FiraSans-Medium-8f9a781e4970d388.woff2,SourceCodePro-Regular-562dcc5011b6de7d.ttf.woff2,SourceCodePro-Semibold-d899c5a5c4aeb14a.ttf.woff2".split(",").map(f=>`<link rel="preload" as="font" type="font/woff2" crossorigin href="../static.files/${f}">`).join(""))</script><link rel="stylesheet" href="../static.files/normalize-76eba96aa4d2e634.css"><link rel="stylesheet" href="../static.files/rustdoc-fecee23daa4e8a6d.css"><meta name="rustdoc-vars" data-root-path="../" data-static-root-path="../static.files/" data-current-crate="hydro_lang" data-themes="" data-resource-suffix="" data-rustdoc-version="1.83.0-nightly (4ac7bcbaa 2024-09-04)" data-channel="nightly" data-search-js="search-a99f1315e7cc5121.js" data-settings-js="settings-4313503d2e1961c2.js" ><script src="../static.files/storage-118b08c4c78b968e.js"></script><script defer src="../crates.js"></script><script defer src="../static.files/main-921df33f47b8780c.js"></script><noscript><link rel="stylesheet" href="../static.files/noscript-3b12f09e550e0385.css"></noscript><link rel="alternate icon" type="image/png" href="../static.files/favicon-32x32-422f7d1d52889060.png"><link rel="icon" type="image/svg+xml" href="../static.files/favicon-2c020d218678b618.svg"></head><body class="rustdoc mod crate"><!--[if lte IE 11]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><nav class="mobile-topbar"><button class="sidebar-menu-toggle" title="show sidebar"></button></nav><nav class="sidebar"><div class="sidebar-crate"><h2><a href="../hydro_lang/index.html">hydro_<wbr>lang</a><span class="version">0.11.0</span></h2></div><div class="sidebar-elems"><ul class="block"><li><a id="all-types" href="all.html">All Items</a></li></ul><section><ul class="block"><li><a href="#reexports">Re-exports</a></li><li><a href="#modules">Modules</a></li><li><a href="#macros">Macros</a></li></ul></section></div></nav><div class="sidebar-resizer"></div><main><div class="width-limiter"><rustdoc-search></rustdoc-search><section id="main-content" class="content"><div class="main-heading"><h1>Crate <a class="mod" href="#">hydro_lang</a><button id="copy-path" title="Copy item path to clipboard">Copy item path</button></h1><span class="out-of-band"><a class="src" href="../src/hydro_lang/lib.rs.html#1-62">source</a> · <button id="toggle-all-docs" title="collapse all docs">[<span>&#x2212;</span>]</button></span></div><h2 id="reexports" class="section-header">Re-exports<a href="#reexports" class="anchor">§</a></h2><ul class="item-table"><li><div class="item-name" id="reexport.RUNTIME_CONTEXT"><code>pub use runtime_context::<a class="static" href="runtime_context/static.RUNTIME_CONTEXT.html" title="static hydro_lang::runtime_context::RUNTIME_CONTEXT">RUNTIME_CONTEXT</a>;</code></div></li><li><div class="item-name" id="reexport.Bounded"><code>pub use boundedness::<a class="enum" href="boundedness/enum.Bounded.html" title="enum hydro_lang::boundedness::Bounded">Bounded</a>;</code></div></li><li><div class="item-name" id="reexport.Unbounded"><code>pub use boundedness::<a class="enum" href="boundedness/enum.Unbounded.html" title="enum hydro_lang::boundedness::Unbounded">Unbounded</a>;</code></div></li><li><div class="item-name" id="reexport.NoOrder"><code>pub use stream::<a class="struct" href="stream/struct.NoOrder.html" title="struct hydro_lang::stream::NoOrder">NoOrder</a>;</code></div></li><li><div class="item-name" id="reexport.Stream"><code>pub use stream::<a class="struct" href="stream/struct.Stream.html" title="struct hydro_lang::stream::Stream">Stream</a>;</code></div></li><li><div class="item-name" id="reexport.TotalOrder"><code>pub use stream::<a class="struct" href="stream/struct.TotalOrder.html" title="struct hydro_lang::stream::TotalOrder">TotalOrder</a>;</code></div></li><li><div class="item-name" id="reexport.Singleton"><code>pub use singleton::<a class="struct" href="singleton/struct.Singleton.html" title="struct hydro_lang::singleton::Singleton">Singleton</a>;</code></div></li><li><div class="item-name" id="reexport.Optional"><code>pub use optional::<a class="struct" href="optional/struct.Optional.html" title="struct hydro_lang::optional::Optional">Optional</a>;</code></div></li><li><div class="item-name" id="reexport.CLUSTER_SELF_ID"><code>pub use location::cluster::<a class="static" href="location/cluster/static.CLUSTER_SELF_ID.html" title="static hydro_lang::location::cluster::CLUSTER_SELF_ID">CLUSTER_SELF_ID</a>;</code></div></li><li><div class="item-name" id="reexport.Cluster"><code>pub use location::<a class="struct" href="location/cluster/struct.Cluster.html" title="struct hydro_lang::location::cluster::Cluster">Cluster</a>;</code></div></li><li><div class="item-name" id="reexport.ClusterId"><code>pub use location::<a class="struct" href="location/cluster/cluster_id/struct.ClusterId.html" title="struct hydro_lang::location::cluster::cluster_id::ClusterId">ClusterId</a>;</code></div></li><li><div class="item-name" id="reexport.ExternalProcess"><code>pub use location::<a class="struct" href="location/external_process/struct.ExternalProcess.html" title="struct hydro_lang::location::external_process::ExternalProcess">ExternalProcess</a>;</code></div></li><li><div class="item-name" id="reexport.Location"><code>pub use location::<a class="trait" href="location/trait.Location.html" title="trait hydro_lang::location::Location">Location</a>;</code></div></li><li><div class="item-name" id="reexport.Process"><code>pub use location::<a class="struct" href="location/process/struct.Process.html" title="struct hydro_lang::location::process::Process">Process</a>;</code></div></li><li><div class="item-name" id="reexport.Tick"><code>pub use location::<a class="struct" href="location/tick/struct.Tick.html" title="struct hydro_lang::location::tick::Tick">Tick</a>;</code></div></li><li><div class="item-name" id="reexport.Timestamped"><code>pub use location::<a class="struct" href="location/tick/struct.Timestamped.html" title="struct hydro_lang::location::tick::Timestamped">Timestamped</a>;</code></div></li><li><div class="item-name" id="reexport.FlowBuilder"><code>pub use builder::<a class="struct" href="builder/struct.FlowBuilder.html" title="struct hydro_lang::builder::FlowBuilder">FlowBuilder</a>;</code></div></li><li><div class="item-name" id="reexport.dfir_rs"><code>pub use <a class="mod" href="../dfir_rs/index.html" title="mod dfir_rs">dfir_rs</a>;</code></div></li></ul><h2 id="modules" class="section-header">Modules<a href="#modules" class="anchor">§</a></h2><ul class="item-table"><li><div class="item-name"><a class="mod" href="boundedness/index.html" title="mod hydro_lang::boundedness">boundedness</a></div></li><li><div class="item-name"><a class="mod" href="builder/index.html" title="mod hydro_lang::builder">builder</a></div></li><li><div class="item-name"><a class="mod" href="cycle/index.html" title="mod hydro_lang::cycle">cycle</a></div></li><li><div class="item-name"><a class="mod" href="deploy/index.html" title="mod hydro_lang::deploy">deploy</a></div></li><li><div class="item-name"><a class="mod" href="deploy_runtime/index.html" title="mod hydro_lang::deploy_runtime">deploy_<wbr>runtime</a></div></li><li><div class="item-name"><a class="mod" href="ir/index.html" title="mod hydro_lang::ir">ir</a></div></li><li><div class="item-name"><a class="mod" href="location/index.html" title="mod hydro_lang::location">location</a></div></li><li><div class="item-name"><a class="mod" href="optional/index.html" title="mod hydro_lang::optional">optional</a></div></li><li><div class="item-name"><a class="mod" href="rewrites/index.html" title="mod hydro_lang::rewrites">rewrites</a></div></li><li><div class="item-name"><a class="mod" href="runtime_context/index.html" title="mod hydro_lang::runtime_context">runtime_<wbr>context</a></div></li><li><div class="item-name"><a class="mod" href="singleton/index.html" title="mod hydro_lang::singleton">singleton</a></div></li><li><div class="item-name"><a class="mod" href="stream/index.html" title="mod hydro_lang::stream">stream</a></div></li></ul><h2 id="macros" class="section-header">Macros<a href="#macros" class="anchor">§</a></h2><ul class="item-table"><li><div class="item-name"><a class="macro" href="macro.q.html" title="macro hydro_lang::q">q</a></div><div class="desc docblock-short">Creates a quoted expression for Hydro.</div></li></ul></section></div></main></body></html>