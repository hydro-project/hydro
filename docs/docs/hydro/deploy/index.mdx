---
sidebar_position: 1
---

import HydroDeployDocs from '../../../../hydro_deploy/core/README.md'

# Hydro Deploy

:::caution

Hydro Deploy is currently in alpha, and documentation is still under construction. Please reach out if you run into any issues--it is under active development and we continue to use it for large-scale distributed experiments within the Hydro research group.

:::

<HydroDeployDocs />

{/* This guide will walk you through the process of deploying your Hydro app in a variety of scenarios. */}
