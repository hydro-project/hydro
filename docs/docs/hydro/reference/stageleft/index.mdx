---
title: Stageleft
sidebar_position: 0
---

Hydro uses a library called Stageleft to power the `q!` macro and code generation logic. 

For detailed documentation on Stageleft's architecture and the `q!` macro, see the [Stageleft README](https://github.com/hydro-project/stageleft/blob/main/README.md).

:::note

Hydro programmers need not understand how Stageleft works under the hood; this information is provided here for readers interested in how its `q!` macro works.

:::
