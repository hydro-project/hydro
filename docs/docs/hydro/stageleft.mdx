---
title: Stageleft
sidebar_position: 10
---

import StageleftDocs from 'stageleft/README.md'

Under the hood, Hydro uses a library called Stageleft to power the `q!` macro and code generation logic. The following docs, from the Stageleft README, outline the core architecture of Stageleft.

:::note

Hydro programmers need not understand how to use Stageleft; this information is provided here for readers
interested in how its `q!` macro works.

:::

<StageleftDocs/>
