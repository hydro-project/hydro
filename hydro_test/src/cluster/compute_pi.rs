use std::time::Duration;

use hydro_lang::*;

pub struct Worker {}
pub struct Leader {}

pub fn compute_pi<'a>(
    flow: &FlowBuilder<'a>,
    batch_size: usize,
) -> (Cluster<'a, Worker>, Process<'a, Leader>) {
    let cluster = flow.cluster();
    let process = flow.process();

    let trials = cluster
        .tick()
        .spin_batch(q!(batch_size))
        .map(q!(|_| rand::random::<(f64, f64)>()))
        .map(q!(|(x, y)| x * x + y * y < 1.0))
        .fold(
            q!(|| (0u64, 0u64)),
            q!(|(inside, total), sample_inside| {
                if sample_inside {
                    *inside += 1;
                }

                *total += 1;
            }),
        )
        .all_ticks();

    let estimate = trials
        .send_bincode_anonymous(&process)
        .reduce_commutative(q!(|(inside, total), (inside_batch, total_batch)| {
            *inside += inside_batch;
            *total += total_batch;
        }));

    unsafe {
        // SAFETY: intentional non-determinism
        estimate.sample_every(q!(Duration::from_secs(1)))
    }
    .for_each(q!(|(inside, total)| {
        println!(
            "pi: {} ({} trials)",
            4.0 * inside as f64 / total as f64,
            total
        );
    }));

    (cluster, process)
}

#[cfg(test)]
mod tests {
    use hydro_lang::Location;
    use hydro_lang::deploy::DeployRuntime;
    use hydro_lang::rewrites::{decoupler, persist_pullup};
    use stageleft::RuntimeData;

    struct DecoupledCluster {}

    #[test]
    fn compute_pi_ir() {
        let builder = hydro_lang::FlowBuilder::new();
        let _ = super::compute_pi(&builder, 8192);
        let built = builder.with_default_optimize::<DeployRuntime>();

        insta::assert_debug_snapshot!(built.ir());

        for (id, ir) in built.compile(&RuntimeData::new("FAKE")).all_dfir() {
            insta::with_settings!({snapshot_suffix => format!("surface_graph_{id}")}, {
                insta::assert_snapshot!(ir.surface_syntax_string());
            });
        }
    }

    #[test]
    fn decoupled_compute_pi_ir() {
        let builder = hydro_lang::FlowBuilder::new();
        let _ = super::compute_pi(&builder, 8192);
        let decoupled_cluster = builder.cluster::<DecoupledCluster>();
        let decoupler = decoupler::Decoupler {
            nodes_to_decouple: vec![4],
            new_location: decoupled_cluster.id().clone(),
        };
        let built = builder
            .optimize_with(persist_pullup::persist_pullup)
            .optimize_with(|leaves| decoupler::decouple(leaves, &decoupler))
            .into_deploy::<DeployRuntime>();

        insta::assert_debug_snapshot!(built.ir());

        for (id, ir) in built.compile(&RuntimeData::new("FAKE")).all_dfir() {
            insta::with_settings!({snapshot_suffix => format!("surface_graph_{id}")}, {
                insta::assert_snapshot!(ir.surface_syntax_string());
            });
        }
    }
}
