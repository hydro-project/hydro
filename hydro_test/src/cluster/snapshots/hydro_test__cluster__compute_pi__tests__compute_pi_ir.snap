---
source: hydro_test/src/cluster/compute_pi.rs
expression: built.ir()
---
[
    ForEach {
        f: stageleft :: runtime_support :: fn1_type_hint :: < (u64 , u64) , () > ({ use crate :: __staged :: cluster :: compute_pi :: * ; | (inside , total) | { println ! ("pi: {} ({} trials)" , 4.0 * inside as f64 / total as f64 , total) ; } }),
        input: Map {
            f: stageleft :: runtime_support :: fn1_type_hint :: < ((u64 , u64) , ()) , (u64 , u64) > ({ use hydro_lang :: __staged :: optional :: * ; | (d , _signal) | d }),
            input: CrossSingleton {
                left: Reduce {
                    f: stageleft :: runtime_support :: fn2_borrow_mut_type_hint :: < (u64 , u64) , (u64 , u64) , () > ({ use crate :: __staged :: cluster :: compute_pi :: * ; | (inside , total) , (inside_batch , total_batch) | { * inside += inside_batch ; * total += total_batch ; } }),
                    input: Persist {
                        inner: Map {
                            f: stageleft :: runtime_support :: fn1_type_hint :: < (hydro_lang :: location :: cluster :: cluster_id :: ClusterId < hydro_test :: cluster :: compute_pi :: Worker > , (u64 , u64)) , (u64 , u64) > ({ use hydro_lang :: __staged :: stream :: * ; | (_ , b) | b }),
                            input: Network {
                                from_key: None,
                                to_location: Process(
                                    1,
                                ),
                                to_key: None,
                                serialize_fn: Some(
                                    :: hydro_lang :: runtime_support :: stageleft :: runtime_support :: fn1_type_hint :: < (u64 , u64) , _ > (| data | { hydro_lang :: runtime_support :: bincode :: serialize (& data) . unwrap () . into () }),
                                ),
                                instantiate_fn: <network instantiate>,
                                deserialize_fn: Some(
                                    | res | { let (id , b) = res . unwrap () ; (hydro_lang :: ClusterId :: < hydro_test :: cluster :: compute_pi :: Worker > :: from_raw (id) , hydro_lang :: runtime_support :: bincode :: deserialize :: < (u64 , u64) > (& b) . unwrap ()) },
                                ),
                                input: Fold {
                                    init: stageleft :: runtime_support :: fn0_type_hint :: < (u64 , u64) > ({ use crate :: __staged :: cluster :: compute_pi :: * ; | | (0u64 , 0u64) }),
                                    acc: stageleft :: runtime_support :: fn2_borrow_mut_type_hint :: < (u64 , u64) , bool , () > ({ use crate :: __staged :: cluster :: compute_pi :: * ; | (inside , total) , sample_inside | { if sample_inside { * inside += 1 ; } * total += 1 ; } }),
                                    input: Map {
                                        f: stageleft :: runtime_support :: fn1_type_hint :: < (f64 , f64) , bool > ({ use crate :: __staged :: cluster :: compute_pi :: * ; | (x , y) | x * x + y * y < 1.0 }),
                                        input: Map {
                                            f: stageleft :: runtime_support :: fn1_type_hint :: < () , (f64 , f64) > ({ use crate :: __staged :: cluster :: compute_pi :: * ; | _ | rand :: random :: < (f64 , f64) > () }),
                                            input: Map {
                                                f: stageleft :: runtime_support :: fn1_type_hint :: < usize , () > ({ use hydro_lang :: __staged :: location :: tick :: * ; | _ | () }),
                                                input: FlatMap {
                                                    f: stageleft :: runtime_support :: fn1_type_hint :: < () , std :: ops :: Range < usize > > ({ use hydro_lang :: __staged :: location :: tick :: * ; let batch_size__free = { use crate :: __staged :: cluster :: compute_pi :: * ; let batch_size__free = 8192usize ; batch_size__free } ; move | _ | 0 .. batch_size__free }),
                                                    input: Source {
                                                        source: Spin,
                                                        location_kind: Cluster(
                                                            0,
                                                        ),
                                                        metadata: HydroIrMetadata {
                                                            location_kind: Cluster(
                                                                0,
                                                            ),
                                                            output_type: Some(
                                                                (),
                                                            ),
                                                            cardinality: None,
                                                            cpu_usage: None,
                                                        },
                                                    },
                                                    metadata: HydroIrMetadata {
                                                        location_kind: Cluster(
                                                            0,
                                                        ),
                                                        output_type: Some(
                                                            usize,
                                                        ),
                                                        cardinality: None,
                                                        cpu_usage: None,
                                                    },
                                                },
                                                metadata: HydroIrMetadata {
                                                    location_kind: Cluster(
                                                        0,
                                                    ),
                                                    output_type: Some(
                                                        (),
                                                    ),
                                                    cardinality: None,
                                                    cpu_usage: None,
                                                },
                                            },
                                            metadata: HydroIrMetadata {
                                                location_kind: Cluster(
                                                    0,
                                                ),
                                                output_type: Some(
                                                    (f64 , f64),
                                                ),
                                                cardinality: None,
                                                cpu_usage: None,
                                            },
                                        },
                                        metadata: HydroIrMetadata {
                                            location_kind: Cluster(
                                                0,
                                            ),
                                            output_type: Some(
                                                bool,
                                            ),
                                            cardinality: None,
                                            cpu_usage: None,
                                        },
                                    },
                                    metadata: HydroIrMetadata {
                                        location_kind: Cluster(
                                            0,
                                        ),
                                        output_type: Some(
                                            (u64 , u64),
                                        ),
                                        cardinality: None,
                                        cpu_usage: None,
                                    },
                                },
                                metadata: HydroIrMetadata {
                                    location_kind: Process(
                                        1,
                                    ),
                                    output_type: Some(
                                        (u64 , u64),
                                    ),
                                    cardinality: None,
                                    cpu_usage: None,
                                },
                            },
                            metadata: HydroIrMetadata {
                                location_kind: Process(
                                    1,
                                ),
                                output_type: Some(
                                    (u64 , u64),
                                ),
                                cardinality: None,
                                cpu_usage: None,
                            },
                        },
                        metadata: HydroIrMetadata {
                            location_kind: Process(
                                1,
                            ),
                            output_type: Some(
                                (u64 , u64),
                            ),
                            cardinality: None,
                            cpu_usage: None,
                        },
                    },
                    metadata: HydroIrMetadata {
                        location_kind: Process(
                            1,
                        ),
                        output_type: Some(
                            (u64 , u64),
                        ),
                        cardinality: None,
                        cpu_usage: None,
                    },
                },
                right: Map {
                    f: stageleft :: runtime_support :: fn1_type_hint :: < tokio :: time :: Instant , () > ({ use hydro_lang :: __staged :: optional :: * ; | _u | () }),
                    input: Source {
                        source: Stream(
                            { use hydro_lang :: __staged :: location :: * ; let interval__free = { use crate :: __staged :: cluster :: compute_pi :: * ; Duration :: from_secs (1) } ; tokio_stream :: wrappers :: IntervalStream :: new (tokio :: time :: interval (interval__free)) },
                        ),
                        location_kind: Process(
                            1,
                        ),
                        metadata: HydroIrMetadata {
                            location_kind: Process(
                                1,
                            ),
                            output_type: Some(
                                tokio :: time :: Instant,
                            ),
                            cardinality: None,
                            cpu_usage: None,
                        },
                    },
                    metadata: HydroIrMetadata {
                        location_kind: Process(
                            1,
                        ),
                        output_type: Some(
                            (),
                        ),
                        cardinality: None,
                        cpu_usage: None,
                    },
                },
                metadata: HydroIrMetadata {
                    location_kind: Process(
                        1,
                    ),
                    output_type: Some(
                        ((u64 , u64) , ()),
                    ),
                    cardinality: None,
                    cpu_usage: None,
                },
            },
            metadata: HydroIrMetadata {
                location_kind: Process(
                    1,
                ),
                output_type: Some(
                    (u64 , u64),
                ),
                cardinality: None,
                cpu_usage: None,
            },
        },
        metadata: HydroIrMetadata {
            location_kind: Process(
                1,
            ),
            output_type: Some(
                (u64 , u64),
            ),
            cardinality: None,
            cpu_usage: None,
        },
    },
]
